### PR TITLE
feat: Add eth_sign friction

### PR DESCRIPTION
**Pull request by NicolasMassart**
## Development & PR Process
1. Follow MetaMask [Mobile Coding Standards](https://docs.google.com/document/d/1VJLwTRsUw_5EDq_o8d6sSbXUAYENLurkRitYO45eY5o/edit?usp=sharing)
2. Add `release-xx` label to identify the PR slated for a upcoming release (will be used in release discussion)
3. Add `needs-dev-review` label when work is completed
4. Add `needs-qa` label when dev review is completed
5. Add `QA Passed` label when QA has signed off

## Description

_1. What is the reason for the change?_

Add friction when enabling the eth_sign RPC endpoint

_2. What is the improvement/solution?_

- Update the setting switch to clarify the risk
- Add a bottom sheet that pops when attempting to switch the value to on
- have a two steps friction on the bottom sheet: 1 - checkbox, 2 - text field to fill (case insensitive, like for "delete wallet" and prevent pasting)
- Enable setting only after all friction steps completed
- add tracking events for analytics

## QA builds

- [iOS install page](https://app.bitrise.io/app/be69d4368ee7e86d/build/850fcc6d-74cf-4c8f-b549-a204c7be715b/artifact/1c2d29c97af09c58/p/7038e0571cc39beea13c7ffeec0cbd7a)
- [Android install page](https://app.bitrise.io/app/be69d4368ee7e86d/build/6e0c1307-3e64-48d9-9cb3-a09f0e7b0cc8/artifact/870a158c4fa9f663/p/508e102029406c1747ec4410bb05301b)

[Other QA Artifacts are available on Bitrise Pipeline build #8122](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/01bf29f9-c91b-4fef-b40f-144f7200f6b1?tab=artifacts)

### Light / Dark modes (on iOS)

| Mode | Advanced setting (off) | CheckBox step | TextInput step (matching) | Advanced setting (on) |
|---|---|---|---|---|
| Light Mode | Light mode Advanced setting (off)| Light mode Checkbox step | Light mode Text field step | Light mode Advanced setting (on) |
| Dark Mode | Dark Mode Advanced setting (off) | Dark Mode Checkbox step | Dark Mode Text field step | Mode Text Advanced setting (on) |

### Screen recording (Light mode on iOS)

https://github.com/MetaMask/metamask-mobile/assets/4677568/ec19f75d-1d3a-46dc-b4dc-43ec4e1bcf9f

## Metrics

metrics event log look like this:

```
 LOG  [MetaMask DEBUG]: Analytics 'trackEvent' - {"category": "Settings Viewed", "properties": {"action": "eth_sign_checkbox_seen"}} {} undefined undefined
 LOG  [MetaMask DEBUG]: Analytics 'trackEvent' - {"category": "Settings Viewed", "properties": {"action": "eth_sign_input_seen"}} {} undefined undefined
 LOG  [MetaMask DEBUG]: Analytics 'trackEvent' - {"category": "Settings Updated", "properties": {"action": "eth_sign_enabled"}} {} undefined undefined
 LOG  [MetaMask DEBUG]: Analytics 'trackEvent' - {"category": "Settings Updated", "properties": {"action": "eth_sign_disabled"}} {} undefined undefined
```
Lines meaning:

1. switch was off and I clicked on i, popping the bottom sheet, viewing the checkbox
2. clicked on continue and moved to second sheet step with text input
3. clicked enabled, it sets the eth_sign setting value to true and closes the bottom sheet
4. then clicked switch off again in the advanced settings without the bottom sheet shown.

looks like the following on MixPanel:
image
image

## Issue

fixes MetaMask/mobile-planning#918

## Checklist

* [x] There is a related GitHub issue
* [x] Tests are included if applicable
* [x] Any added code is fully documented

### CODE REVIEW

File: app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+      fontSize: 20,
+      textAlign: 'center',
+      lineHeight: breakPoint ? 24 : 26,
+    },
+    red: {
+      marginHorizontal: 24,
+      color: colors.error.default,
+    },
+    warning: {
+      ...fontStyles.normal,
```
Comment (Cal-L) [inline app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:43]: Move typography styles to `warningText` and replace with `sBodyMD` instead of hardcoded values

File: app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      display: 'flex',
+      flexDirection: 'row',
+      padding: 8,
+    },
+    warningText: {
+      lineHeight: breakPoint ? 18 : 22,
+      flex: 1,
+    },
+    bold: {
+      ...fontStyles.bold,
```
Comment (Cal-L) [inline app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:62]: Replace with `sBodyMDBold` from typography theme

File: app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    },
+    understandCheckText: {
+      marginEnd: 16,
+    },
+    understandCheckboxView: {
+      margin: 16,
+      flexDirection: 'row',
+      alignItems: 'stretch',
+      justifyContent: 'flex-start',
+      display: 'flex',
```
Comment (Cal-L) [inline app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:75]: Remove `display: flex`

File: app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      fontSize: 14,
```
Comment (Cal-L) [inline app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:76]: Remove font style from View

File: app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      flex: 1,
+    },
+    bold: {
+      ...fontStyles.bold,
+    },
+    understandCheckbox: {
+      marginEnd: 8,
+    },
+    understandCheckText: {
+      marginEnd: 16,
```
Comment (Cal-L) [inline app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:68]: Missing typography `sBodyMD`

File: app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    understandCheckboxView: {
+      margin: 16,
+      flexDirection: 'row',
+      alignItems: 'stretch',
+      justifyContent: 'flex-start',
+      display: 'flex',
+      fontSize: 14,
+    },
+    textConfirmField: {
+      fontSize: 14,
```
Comment (Cal-L) [inline app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:79]: Replace font styles with `sBodyMDBold` and provide `color` property

File: app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts (hunk 5 above)
Comment (Cal-L) [reply app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:68]: + provide `color` property

File: app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+      textAlign: 'left',
+    },
+    textConfirmWarningText: {
+      fontSize: 12,
```
Comment (Cal-L) [inline app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:83]: Provide `color` property and use `sBodySM`

File: app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    },
+    buttonsContainer: {
+      flexDirection: 'row',
+      display: 'flex',
```
Comment (Cal-L) [inline app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:87]: Remove display flex

File: app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    },
+    buttonStart: {
+      flex: 1,
+      marginEnd: 8,
```
Comment (Cal-L) [inline app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:91]: remove `marginEnd` here and apply `marginHorizontal` to `buttonsContainer`

File: app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    },
+    buttonEnd: {
+      flex: 1,
+      marginStart: 8,
```
Comment (Cal-L) [inline app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:95]: remove `marginEnd` here and apply `marginHorizontal` to `buttonsContainer`

File: app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    },
+    areYouSure: {
+      width: '100%',
+      paddingVertical: breakPoint ? 16 : 24,
+      justifyContent: 'center',
+      alignSelf: 'center',
```
Comment (Cal-L) [inline app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:101]: This section should take up the width completely with horizontal margins and the content should be aligned to the left (by default). `alignSelf: center` is centering it at the moment.

File: app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts [hunk 12]
```diff
@@ -0,0 +1,103 @@
+// Third party dependencies.
+import { StyleSheet } from 'react-native';
+import { fontStyles } from '../../../../../styles/common';
+import Device from '../../../../../util/device';
+
+const breakPoint = Device.getDeviceHeight() < 700;
```
Comment (Cal-L) [inline app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:6]: Why is this needed and does it break the UI if it isn't included?

File: app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+
+/**
+ * Style sheet function for EthSignFriction component.
+ *
+ * @returns StyleSheet object.
+ */
+
+export default (colors: any) =>
```
Comment (Cal-L) [inline app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:14]: Type this with `import { ThemeColors } from '@metamask/design-tokens/dist/js/themes/types';`

**Review by Cal-L — COMMENTED** [review]
Left some comments

**Comment by Cal-L** [comment]
Please also run the app against dark mode since it will show you the missing color properties that need to be applied.

File: app/component-library/components/Sheet/SheetBottom/SheetBottom.tsx [hunk 14]
```diff
@@ -233,7 +235,12 @@ const SheetBottom = forwardRef<SheetBottomRef, SheetBottomProps>(
     const renderNotch = () => isInteractable && <View style={styles.notch} />;
 
     return (
-      <View style={styles.base} {...props}>
+      <KeyboardAvoidingView
+        behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
+        keyboardVerticalOffset={Platform.OS === 'ios' ? -20 : 64}
```
[the remark this answers is not in the document]
Comment (NicolasMassart) [reply app/component-library/components/Sheet/SheetBottom/SheetBottom.tsx:240]: screenBottomPadding is too much on iOS and not enough on Android. Any idea how to fix this?

[the remark this answers is not in the document]
Comment (NicolasMassart) [reply app/component-library/components/Sheet/SheetBottom/SheetBottom.tsx:240]: screenBottomPadding value is 34 on iOS and 0 on Android which is not close to my need.

File: app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts (hunk 9 above)
Comment (NicolasMassart) [reply app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:91]: Any idea how I can create the gap between the two buttons without this as we don't have access to the `gap` flex yet in the current RN version?

File: app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts (hunk 5 above)
Comment (NicolasMassart) [reply app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts]: Why not the default color?

**Review by Cal-L — DISMISSED** [review]
LGTM

**Review by Cal-L — APPROVED** [review]
Lgtm

**Comment by seaona** [comment]
A couple of findings/comments from QA, the rest looks good :100: 

## Sending ENTER twice triggers other functionalities (i.e. Reset Account function, IPFS Gateway..)
Whenever I input the text and I submit ENTER key, I see that the the first one validates the text, but the second one, instead of `Enabling` the function it triggers another action, depending on where I am on the background screen. This could be Reset Account, Show Hex Data, IPFS Gateway etc..


https://github.com/MetaMask/metamask-mobile/assets/54408225/f8910ad5-5f96-4139-9eb0-038a1ea18710

## `Enable` button is not enabled unless I submit the text
This I am not sure if it's an issue, but commenting just in case. Whenever I enter the correct text, I can see how the button is not enabled, unless I submit (ENTER)


https://github.com/MetaMask/metamask-mobile/assets/54408225/bbe40bf9-7baa-4e48-9dcf-898737aac2a7

## Bold text
I can see how the ON/OFF text is not in bold in Android. Not sure if this is expected

image

**Comment by NicolasMassart** [comment]
thanks @seaona, I will investigate the first issue "Sending ENTER twice triggers".

For the second one, I'm going to confirm with design team what the expected behaviour is as this is what I understood from their Figma design (link in the issue)

For the third one, it doesn't look very bold but it is 😅 this is the `sBodyLGMedium` style as indicated in the design.
image

**Comment by seaona** [comment]
thank you for your comments @NicolasMassart @hesterbruikman ! A summary of our discussions and outcomes:
 - **Issue 1**: would be tackled in a separate PR. The issue is here https://github.com/MetaMask/metamask-mobile/issues/6613
 - **Issue 2**: seems that input validation behavior is different in several places, so this is not a blocker and can be consolidated on a separate PR
 - **Issue 3**: not an issue as per @NicolasMassart 's comment
Given that, PR looks good from QA side
